File: 프로그래머스/프로세스.py

from collections import deque


def solution(priorities, location):
    queue = [(i, p) for i, p in enumerate(priorities)]  # index와 인덱스 쌍으로 이뤄진 튜플로 변환해 queue에 저장.
    answer = 0
    while True:
        cur = queue.pop(0)  # queue의 첫번째 요소 추출

        if any(cur[1] < q[1] for q in queue):  # 현재 우선순위 보다 더 높은 우선순위가 있다면
            queue.append(cur)  # queue의 뒤쪽에 넣는다.
        else:  # 현재 우선순위가 가장 높은 경우
            answer += 1
            if cur[0] == location:  # 현재 요소의 인덱스와 location가 일치할 경우
                return answer


print(solution([2, 1, 3, 2], 2))
print(solution([1, 1, 9, 1, 1, 1], 0))

File: 프로그래머스/2/42587. 프로세스/프로세스.java
import java.util.*;

class Solution {
    public int solution(int[] priorities, int location) {
        int answer = 0;
        
        // 우선순위 큐 생성: 우선순위 내림차순 정렬
        PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());
        
        for(int pri: priorities){
            pq.add(pri); // 우선순위 큐에 추가
        }
        
        while(!pq.isEmpty()){
            for(int i=0; i<priorities.length; i++){
                if(priorities[i]==pq.peek()){
                    pq.poll();
                    answer ++; // 순서 증가
                    
                    if(i== location){ // 구하려는 위치라면
                        return answer;
                    }
                }
            }
        }
        
        return answer;
    }
}